### PR TITLE
Polycube CI with GitHub Actions

### PR DESCRIPTION
**Pull request by klaalk**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+                          "title":"Commit",
+                          "value":`${process.env.AS_COMMIT}`,
+                        }]
+                  }]
+              }
+        env:
+          SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
+        if: always() # Pick up events even if the job fails or is canceled.
+
+  # This job deal with building the application
```
Comment (frisso) [inline .github/workflows/ci.yml:92]:
```suggestion
  # This job deals with building the application
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+      - name: Image digest
+        run: echo ${{ steps.docker_build.outputs.digest }}
+
+      - name: Save build result in output
+        id: save
+        if: always()
+        run: |
+          echo "::set-output name=${{ matrix.mode }}::${{ job.status }}"
+  
+  # This job deal with tests
```
Comment (frisso) [inline .github/workflows/ci.yml:180]:
```suggestion
  # This job deals with tests
```

**Review by frisso — APPROVED** [review]

File: Documentation/developers/polycube-ci.rst [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+The first event is used to build and test any open PRs, then the workflow assumes a "dev" state for its entire duration.
+
+The second workflow is used to build and push an updated version of polycube on docker after a PR merge. In this situation the status of the workflow assumes the value of "master".
+
+The third event is used to build and push a polycube release on docker with an official tag. The status of the workflow becomes "release" and a changelog is also published on github starting from the last release.
+
+
+Workflow in "dev" state
+-------------------------
+This workflow is used to validate any open PR for this reason all tests are performed after the build. If there is only one failed test, the entire workflow will end in error, reporting it via annotation. Furthermore, it is possible to find the logs of the tests through the aritfact section within the page of the performed workflow.
```
Comment (FedeParola) [inline Documentation/developers/polycube-ci.rst:35]:
```suggestion
This workflow is used to validate any open PR for this reason all tests are performed after the build. If there is only one failed test, the entire workflow will end in error, reporting it via annotation. Furthermore, it is possible to find the logs of the tests through the artifact section within the page of the performed workflow.
```

File: Documentation/developers/polycube-ci.rst [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+In this state only polycube in default will be built and published on docker under the "polycubenets" repository with it development names "polycube-pr".
```
Comment (FedeParola) [inline Documentation/developers/polycube-ci.rst:36]:
```suggestion
In this state only the full polycube image will be built and published on **Docker Hub** under the ``polycubenets/polycube-pr`` repository.
```

File: Documentation/developers/polycube-ci.rst [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+
+Workflow in "master" state
+----------------------------
+Once validated and accepted a PR this workflow start in order to updated the latest version of polycube by publishing the new image on "polycubenets" repository.
```
Comment (FedeParola) [inline Documentation/developers/polycube-ci.rst:41]:
```suggestion
Once validated and accepted a PR this workflow start in order to updated the latest version of polycube by publishing the new image on ``polycubenets`` repository.
```

File: Documentation/developers/polycube-ci.rst [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+Here the individual names change to polycube, polycube-iptables and polycube-k8s.
```
Comment (FedeParola) [inline Documentation/developers/polycube-ci.rst:42]: Three different images are published: ``polycube``, the full polycube image; ``polycube-iptables``, containing only the files needed to run the polycube clone of iptables; and ``polycube-k8s``, containing only the CNI plugin for Kubernetes.

File: Documentation/developers/polycube-ci.rst [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+
+Workflow in "release" state
+-----------------------------
+This workflow is used to release an official polycube tag. The new image will always be built and published on the "polycubenets" repository with the official names (polycube, polycube-iptables and polycube-k8s).
```
Comment (FedeParola) [inline Documentation/developers/polycube-ci.rst:47]:
```suggestion
This workflow is used to release an official polycube tag. The new image will always be built and published on the ``polycubenets`` repository with the official names (polycube, polycube-iptables and polycube-k8s).
```

**Review by FedeParola — CHANGES_REQUESTED** [review]

**Review by FedeParola — APPROVED** [review]

File: Documentation/developers/polycube-ci.rst [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+Workflow in "dev" state
+-------------------------
+This workflow is used to validate any open PR for this reason all tests are performed after the build. If there is only one failed test, the entire workflow will end in error, reporting it via annotation. Furthermore, it is possible to find the logs of the tests through the artifact section within the page of the performed workflow.
+In this state only the full polycube image will be built and published on **Docker Hub** under the ``polycubenets/polycube-pr`` repository.
+
+
+Workflow in "master" state
+----------------------------
+Once validated and accepted a PR this workflow start in order to updated the latest version of polycube by publishing the new image on ``polycubenets`` repository.
```
Comment (frisso) [inline Documentation/developers/polycube-ci.rst:41]:
```suggestion
This workflow starts once a PR is validated and accepted and it updates the _latest_ version of polycube by publishing the new image on ``polycubenets`` repository.
```

File: Documentation/developers/codegen.rst [hunk 9]
```diff
@@ -9,20 +9,20 @@ It is composed by ``pyang``, that parses the YANG datamodel and creates an inter
 Among the services that are automatically provided, the C++ skeleton handles the case in which a complex object is requested, such as the entire configuration of the service, which is returned by disaggregating the big request into a set of smaller requests for each *leaf* object.
 Hence, this leaves to the programmer only the responsibility to implement the interaction with *leaf* objects.
 
-The easiest way to generate a stub is to use the ``polycubenetwork/polycube-codegen`` docker image.
+The easiest way to generate a stub is to use the ``polycubenets/polycube-codegen`` docker image.
 This contains ``pyang``, ``swagger-codegen`` and a script to invoke them.
 
 In order to run the image, it is needed to mount a volume into ``/polycube-base-datamodels`` where the base datamodels are located on the host (by default ``/usr/local/include/polycube/datamodel-common/``), it is also needed to mount a volume to share the input yang model and the output folder.
```
Comment (frisso) [inline Documentation/developers/codegen.rst:15]:
```suggestion
In order to run the image, you need to mount a volume into ``/polycube-base-datamodels`` where the base datamodels are located on the host (by default ``/usr/local/include/polycube/datamodel-common/``).
Furthermore, you need also to mount a volume to share the input yang model and the output folder.
```

File: Documentation/developers/polycube-ci.rst [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+The first event is used to build and test any open PRs, then the workflow assumes a "dev" state for its entire duration.
+
+The second workflow is used to build and push an updated version of polycube on docker after a PR merge. In this situation the status of the workflow assumes the value of "master".
+
+The third event is used to build and push a polycube release on docker with an official tag. The status of the workflow becomes "release" and a changelog is also published on github starting from the last release.
+
+
+Workflow in "dev" state
+-------------------------
+This workflow is used to validate any open PR for this reason all tests are performed after the build. If there is only one failed test, the entire workflow will end in error, reporting it via annotation. Furthermore, it is possible to find the logs of the tests through the artifact section within the page of the performed workflow.
```
Comment (frisso) [inline Documentation/developers/polycube-ci.rst:35]:
```suggestion
This workflow is used to validate any open PR. For this reason, all tests are performed after the build. If there is only one failed test, the entire workflow will end in error, reporting it via annotation. Furthermore, it is possible to find the logs of the tests through the artifact section within the page of the performed workflow.
```

File: Documentation/developers/polycube-ci.rst [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+The workflow can be used to build, test, package, release, and deploy a project on GitHub.
+
+
+Implementation with polycube
+----------------------------
+You can find the workflow for CI/CD under .github/workflows/ci.yml.
+
+This workflow can start with three different events for three different tasks:
+
+  1.  PR opened/synchronized/reopened
```
Comment (frisso) [inline Documentation/developers/polycube-ci.rst:22]:
```suggestion
  1.  **Pull Request (PR) opened/synchronized/reopened**: used to build and test any open PRs. The workflow assumes a "dev" state for its entire duration.
```

File: Documentation/developers/polycube-ci.rst [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+  2.  Commit push on master (PR accepted)
```
Comment (frisso) [inline Documentation/developers/polycube-ci.rst:23]:
```suggestion
  2.  **Commit push on master (PR accepted)**: used to build and push an updated version of polycube on the public docker registry after a PR merge. In this situation the status of the workflow assumes the value of "master".
```

File: Documentation/developers/polycube-ci.rst [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+  3.  Tag push on master (tag with format "v*")
```
Comment (frisso) [inline Documentation/developers/polycube-ci.rst:24]:
```suggestion
  3.  **Tag push on master (tag with format "v*")**: used to build and push a polycube release on the public docker registry with an official tag. The status of the workflow becomes "release" and a changelog is also published on github starting from the last release.
```

File: Documentation/developers/polycube-ci.rst [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+The first event is used to build and test any open PRs, then the workflow assumes a "dev" state for its entire duration.
```
Comment (frisso) [inline Documentation/developers/polycube-ci.rst:26]:
```suggestion
```

File: Documentation/developers/polycube-ci.rst [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+
+The second workflow is used to build and push an updated version of polycube on docker after a PR merge. In this situation the status of the workflow assumes the value of "master".
```
Comment (frisso) [inline Documentation/developers/polycube-ci.rst:28]:
```suggestion
```

File: Documentation/developers/polycube-ci.rst [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+
+The third event is used to build and push a polycube release on docker with an official tag. The status of the workflow becomes "release" and a changelog is also published on github starting from the last release.
```
Comment (frisso) [inline Documentation/developers/polycube-ci.rst:30]:
```suggestion
```

**Comment by FedeParola** [comment]
Can the merge be forced? Jenkins tests won't pass (since the system doesn't exist anymore). I don't have the privilege to merge

**Pull request merged**